File: samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/play/infrastructure/zlink/spots/tictactoegamespot/handlers/PlayActorLeaveGameHandler.java
package systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.spots.tictactoegamespot.handlers;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.framework.handlers.ZLinkSpotActorSend;
import systems.zlink.samples.tictactoe.server.configuration.SampleNames;
import systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.actors.PlayActor;
import systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.spots.tictactoegamespot.TicTacToeGame;
import systems.zlink.samples.tictactoe.shared.contracts.LeaveGameMsg;

import java.util.concurrent.CompletionStage;

@ZLinkHandlerGroup(SampleNames.PlayActor)
public final class PlayActorLeaveGameHandler {
    private static final Logger LOGGER = LoggerFactory.getLogger(PlayActorLeaveGameHandler.class);

    @ZLinkSpotActorSend
    public CompletionStage<Void> leaveGame(
            TicTacToeGame spot,
            PlayActor actor,
            ZLinkMessageContext context,
            LeaveGameMsg message) {
        LOGGER.info(
                "actor: LeaveGameMsg received. actor={}, roomId={}",
                actor.actorId(),
                message.roomId());
        return spot.leaveGame(actor, message.roomId())
                .thenRun(
                        () ->
                                LOGGER.info(
                                        "tictactoe-lifecycle leave-completed actor={}",
                                        actor.actorId()));
    }
}

File: samples/kotlin/TicTacToe/Server/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/play/infrastructure/zlink/spots/tictactoegamespot/handlers/PlayActorLeaveGameHandler.kt
package systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.spots.tictactoegamespot.handlers

import org.slf4j.LoggerFactory
import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.framework.handlers.ZLinkSpotActorSend
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleNames
import systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.actors.PlayActor
import systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.spots.tictactoegamespot.TicTacToeGame
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.LeaveGameMsg

@ZLinkHandlerGroup(SampleNames.PlayActor)
class PlayActorLeaveGameHandler {
    @ZLinkSpotActorSend
    suspend fun leaveGame(
        spot: TicTacToeGame,
        actor: PlayActor,
        context: ZLinkMessageContext,
        message: LeaveGameMsg,
    ) {
        logger.info(
            "actor: LeaveGameMsg received. actor={}, roomId={}",
            actor.actorId,
            message.roomId,
        )
        spot.leaveGame(actor, message.roomId)
        logger.info("tictactoe-lifecycle leave-completed actor={}", actor.actorId)
    }

    private companion object {
        val logger = LoggerFactory.getLogger(PlayActorLeaveGameHandler::class.java)
    }
}
